engine/flink/scala-utils/src/main/scala/pl/touk/nussknacker/engine/flink/api/typeinfo/caseclass/CaseClassTypeInfoFactory.scala

```diff
@@ -17,8 +17,8 @@ abstract class CaseClassTypeInfoFactory[T <: Product: ClassTag] extends TypeInfo
       genericParameters: java.util.Map[String, TypeInformation[_]]
   ): TypeInformation[T] = {
     val tClass     = classTag[T].runtimeClass.asInstanceOf[Class[T]]
-    val fieldNames = tClass.getDeclaredFields.map(_.getName).toList
-    val fieldTypes = tClass.getDeclaredFields.map(_.getType).map(TypeExtractor.getForClass(_))
+    val fieldNames = tClass.getDeclaredConstructors.head.getParameters.map(_.getName).toList
+    val fieldTypes = tClass.getDeclaredConstructors.head.getParameterTypes.map(TypeExtractor.getForClass(_))
```

Comment:
are we sure that the first constructor on the list is the primary one? I wonder if cannot be different on different JVM implementations. Maybe there is a way to check it?

Reply:
Before implementing it this way i locally checked that the first one should be the primary one.
Unfortunately there does not seem to be an easy way to check it. 
We have to use Java reflection classes here and Java doesn't know anything about  case classes' primary constructors.
I tried mixing Scala reflection to get what we need and ended up with this...
```
abstract class CaseClassTypeInfoFactory[T <: Product: ClassTag] extends TypeInfoFactory[T] with Serializable {

  override def createTypeInfo(
                               t: Type,
                               genericParameters: java.util.Map[String, TypeInformation[_]]
                             ): TypeInformation[T] = {
    val mirror = runtimeMirror(getClass.getClassLoader)
    val classType = classTag[T].runtimeClass
    val primaryConstructor = mirror.classSymbol(classType).primaryConstructor.asMethod
    val fieldNames = primaryConstructor.paramLists.flatten.map(_.name.toString)
    val fieldTypes = primaryConstructor.paramLists.flatten.map(_.typeSignature.getClass).map(TypeExtractor.getForClass(_))

    new CaseClassTypeInfo[T](classType.asInstanceOf[Class[T]], Array.empty, fieldTypes.toIndexedSeq, fieldNames) {
      override def createSerializer(config: ExecutionConfig): TypeSerializer[T] = {
        new ScalaCaseClassSerializer[T](
          classType.asInstanceOf[Class[T]],
          fieldTypes.map(typeInfo => NullableSerializer.wrap(typeInfo.createSerializer(config), true)).toArray
        )
      }
    }
  }

}
```
It passes the tests but looks horrible... With Scala reflection we can get the primary constructor with the `primaryConstructor` call.

Reply:
Well... After consulting the approach with Java and Scala reflection mixture I added a new proposed solution. It's a bit different from the one above as `fieldTypes` didn't contain correct types' sigantures. I also added a test with additional secondary constructor.